File: leetcode/python/0002. Add Two Numbers.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# Time complexity: O(N)
# Space complexity: O(N)

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        res = ListNode()
        temp = res

        carry = 0

        while l1 is not None and l2 is not None:
            curr_value = l1.val + l2.val + carry
            carry = 0
            temp.next = ListNode(val=(curr_value % 10))

            if curr_value >= 10:
                carry = 1

            temp = temp.next

            l1 = l1.next
            l2 = l2.next

        while l1 is not None:
            curr_value = l1.val + carry
            carry = 0
            temp.next = ListNode(val=(curr_value % 10))
            if curr_value >= 10:
                carry = 1
            temp = temp.next
            l1 = l1.next

        while l2 is not None:
            curr_value = l2.val + carry
            carry = 0
            temp.next = ListNode(val=(curr_value % 10))
            if curr_value >= 10:
                carry = 1
            temp = temp.next
            l2 = l2.next

        if carry > 0:
            temp.next = ListNode(val=1)

        return res.next

File: leetcode/java/0002. Add Two Numbers.java
package leetcode.java;

// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

// Time complexity: O(N)
// Space complexity: O(N)

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode res = new ListNode();
        ListNode temp = res;

        int carry = 0;

        while(l1 != null || l2 != null) {

            int sum = carry;

            if (l1 != null) {
                sum += l1.val;
                l1 =l1.next;
            }

            if (l2 != null) {
                sum += l2.val;
                l2 = l2.next;
            }

            temp.next = new ListNode(sum%10);
            temp = temp.next;
            carry = sum / 10;
        }

        if (carry > 0) {
            temp.next = new ListNode(carry);
            temp = temp.next;
        }

        return res.next;
    }
}